cranelift/frontend/src/ssa.rs

```diff
@@ -648,19 +659,9 @@ impl SSABuilder {
                 // so we don't need to have it as a block argument.
                 // We need to replace all the occurrences of val with pred_val but since
                 // we can't afford a re-writing pass right now we just declare an alias.
-                // Resolve aliases eagerly so that we can check for cyclic aliasing,
-                // which can occur in unreachable code.
-                let mut resolved = func.dfg.resolve_aliases(pred_val);
-                if sentinel == resolved {
-                    // Cycle detected. Break it by creating a zero value.
```

Comment:
Why isn't this reachable anymore?

Reply:
Good question. Two pieces:

- `ZeroOneOrMore::One(pred_val)` comes from the above `for pred_val in self` loop, which already resolved aliases. So it's guaranteed at this point that resolving aliases is a no-op.
- The above loop filters out any `pred_val` which is equal to `sentinel`.

So instead of being a special case of `ZeroOneOrMore::One`, this is now just another case of `ZeroOneOrMore::Zero`. And that already had the `emit_zero`.

But that case also takes care of two other things that weren't handled here: appending the current block to the layout if necessary, and adding this block to the side effects. So this does change behavior.

I _think_ it was a bug that this case didn't also do those things. I also think it doesn't matter very much: this can only happen in unreachable blocks. As far as I understand, cranelift-wasm never generates unreachable blocks; does cg-clif?

Reply:
Thanks for the explanation! cg_clif does generate unreachable blocks in some cases.

Reply:
I think this is worth noting in the git history so I'm adding this paragraph to the commit message:

This also merges a special case, where there's exactly one unique non-sentinel definition but it's actually an alias for the sentinel, into the general case where all definitions are from the sentinel. As a result there's only one case that has to introduce a definition of the variable to zero.